### PR TITLE
stitch: Integrate invariant checking into stitch

### PR DESCRIPTION
**Pull request by akshayknarayan**
Add a stitch builtin, "invariant", which inlines invariant checking. To
make this work, now that labels can be queried independently,
modify graph construction to add nodes for all labels,
rather than inferring the presence of a node from
the list of connections in the stitch.

### CODE REVIEW

**Comment by akshayknarayan** [comment]
Note: I didn't migrate the "query" ability since it's unclear how that should be incorporated.

File: stitch/inspect.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	for _, m := range spec.QueryMachines() {
+		g.machines = append(g.machines, getMachine(m))
+	}
+
+	return g, nil
+}
+
+// InspectMain is the main function for inspect tool. Helps visualize stitches.
+func InspectMain(opts []string) {
+	configPath := opts[1]
```
Comment (sirspinach) [inline stitch/inspect.go]: I think this line isn't failing gracefully:

```
aetherboard:quilt shwangster$ quilt inspect
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0xa13d20, 0xc820010090)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/NetSys/quilt/stitch.InspectMain(0xc82000a210, 0x1, 0x1)
        /Users/shwangster/go/src/github.com/NetSys/quilt/stitch/inspect.go:58 +0x91b
main.main()
        /Users/shwangster/go/src/github.com/NetSys/quilt/quilt.go:63 +0x59e
```

Comment (sirspinach) [inline stitch/inspect.go]: This probably should call `usage()` if `len(opts) == 2`.

File: stitch/inspect.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	f, err := os.Open(configPath)
+	if err != nil {
+		panic(err)
```
Comment (sirspinach) [inline stitch/inspect.go]: I'm not sure if a panic message and stacktrace is appropriate for mistyping a file name. Maybe just throw an error instead?

```
aetherboard:quilt shwangster$ quilt inspect a
panic: open a: no such file or directory

goroutine 1 [running]:
panic(0x9c8120, 0xc820210150)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/NetSys/quilt/stitch.InspectMain(0xc8200621c0, 0x2, 0x2)
        /Users/shwangster/go/src/github.com/NetSys/quilt/stitch/inspect.go:62 +0xaf
main.main()
        /Users/shwangster/go/src/github.com/NetSys/quilt/quilt.go:63 +0x59e
```

File: stitch/inspect.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+const quiltPath = "QUILT_PATH"
+
+type argOption struct {
+	result interface{} // Optionally store result.
+}
+
+func usage() {
+	fmt.Fprintln(os.Stderr, "Usage: inspect <path to spec file> [commands]\n"+
```
Comment (sirspinach) [inline stitch/inspect.go]: Perhaps it would be a good idea to explain here what `inspect` does. (I'm still not sure myself haha)

**Comment by sirspinach** [comment]
ok, I'm done reviewing.

File: stitch/inspect.go (hunk 2 above)
Comment (akshayknarayan) [inline stitch/inspect.go]: good catch, I messed up a rebase.

File: stitch/inspect.go (hunk 3 above)
Comment (akshayknarayan) [inline stitch/inspect.go]: good idea, will do

File: stitch/inspect.go (hunk 1 above)
Comment (akshayknarayan) [inline stitch/inspect.go]: good catch, I messed up a rebase.

File: stitch/inspect.go (hunk 2 above)
Comment (sirspinach) [inline stitch/inspect.go]: By throw an error I meant to say 'print an error message'.

Comment (ejj) [inline stitch/inspect.go]: Btw, I generally agree.  Not a big fan of `panic`ing except when it's used more like an assert fail.  I.e. checking for bugs in the code.

**Comment by ejj** [comment]
Normally, I'm not that strict on code coverage, but in the dsl, I've been aiming for 100%.  It's fairly easy to test and bugs can be quite subtle.  Could you up the coverage for the changes you made in stitch?

**Comment by akshayknarayan** [comment]
incorporated @sirspinach 's comments into the first commit. Thanks!

The commit to remove a machine is because stitch already defines a Machine struct with the same information. It's a separate commit because I want to make it easy to revert later if we decide static analysis needs its own way of tracking machines.

The commit to move the visualization tools into a separate package rectifies the weird situation where visualization had to be in the stitch package to access non-exported fields. The relevant fields were exported and visualization was moved (back to the old "inspect" package).

**Comment by akshayknarayan** [comment]
Changes on master have obsoleted this patch, and rebasing has led to significant changes. Closing this PR, will open a new PR for the rebased branch.

**Pull request closed without merging**